src/main/java/picard/illumina/parser/readers/CbclReader.java

```diff
@@ -57,117 +63,189 @@
 
-    public CbclReader(final List<File> cbcls, final int[] outputLengths) {
+    public CbclReader(final List<File> cbcls, final Map<Integer, File> filterFileMap, final int[] outputLengths,
+                      int tileNum, List<AbstractIlluminaPositionFileReader.PositionInfo> locs, boolean headerOnly) {
         super(outputLengths);
+        surfaceToTileToCbclMap = sortCbcls(cbcls);
+        this.filterFileMap = filterFileMap;
         cycleData = new CycleData[cycles];
-        cachedTile = new ByteBuffer[cycles];
-        currentTile = new int[cycles];
+        cachedTile = new byte[cycles][];
+        cachedTilePosition = new int[cycles];
+        for (int i = 1; i <= cycles; i++) {
+            allTiles.put(i, new ArrayList<>());
+        }
+        readSurfaceTile(tileNum, locs, headerOnly);
+        close();
+    }
 
+    private void readSurfaceTile(int tileNum, List<AbstractIlluminaPositionFileReader.PositionInfo> locs,
+                                 boolean headerOnly) {
+        log.info("Processing tile " + tileNum);
         try {
-            final ByteBuffer byteBuffer = ByteBuffer.allocate(INITIAL_HEADER_SIZE);
-            byteBuffer.order(ByteOrder.LITTLE_ENDIAN);
-            for (int i = 0; i < cycles; i++) {
-                currentTile[i] = 0;
-                final File bclFile = cbcls.get(i);
-
-                final InputStream stream = open(bclFile, false, false, false);
-                int read = stream.read(byteBuffer.array());
-
-                //we need to read the first 6 bytes to determine the header size
-                if (read != INITIAL_HEADER_SIZE) {
-                    close();
-                    throw new RuntimeIOException(String.format("BCL %s has invalid header structure.", bclFile.getAbsoluteFile()));
-                }
+            for (Map.Entry<Integer, Map<Integer, File>> entry : surfaceToTileToCbclMap.entrySet()) {
+                Map<Integer, File> cycleMap = entry.getValue();
+                for (int i = 0; i < cycles; i++) {
+                    //cycleMap is 1 indexed
+                    final ByteBuffer byteBuffer = ByteBuffer.allocate(INITIAL_HEADER_SIZE);
+                    byteBuffer.order(ByteOrder.LITTLE_ENDIAN);
+
+                    final File bclFile = cycleMap.get(i + 1);
+                    if (bclFile == null) {
+                        throw new PicardException("Expected cbcl file for surface " + entry.getKey() + " cycle " + (i + 1) + " but it was not found.");
+                    }
+
+                    final InputStream stream = open(bclFile, false, false, false);
+                    int read = stream.read(byteBuffer.array());
 
-                short version = byteBuffer.getShort();
-                int headerSize = byteBuffer.getInt();
+                    //we need to read the first 6 bytes to determine the header size
+                    if (read != INITIAL_HEADER_SIZE) {
+                        close();
```

Comment:
There's a few places where you call `close()` before returning. It would be easier to use `try/finally` around the call to `readSurfaceTile()` to ensure that `close()` is called no matter how the operation fails.